state-chain/runtime/src/chainflip/solana_elections.rs

```diff
@@ -143,28 +146,49 @@ impl OnCheckComplete<<Runtime as Chainflip>::ValidatorId> for OnCheckCompleteHoo
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode, TypeInfo)]
 pub struct TransactionSuccessDetails {
 	pub tx_fee: u64,
+	// It is possible for a contract call to be reverted due to contract's internal error.
+	// This field is set to `true` if the contract call executed successfully without error.
+	pub contract_call_successful: bool,
 }
 
 pub struct SolanaEgressWitnessingHook;
 
 impl OnEgressSuccess<SolSignature, TransactionSuccessDetails> for SolanaEgressWitnessingHook {
 	fn on_egress_success(
 		signature: SolSignature,
-		TransactionSuccessDetails { tx_fee }: TransactionSuccessDetails,
+		TransactionSuccessDetails { tx_fee, contract_call_successful }: TransactionSuccessDetails,
 	) {
 		use cf_traits::KeyProvider;
-
-		if let Err(err) = SolanaBroadcaster::egress_success(
-			pallet_cf_witnesser::RawOrigin::CurrentEpochWitnessThreshold.into(),
-			signature,
-			// Assign any owed fees to the current key.
-			SolanaThresholdSigner::active_epoch_key().map(|e| e.key).unwrap_or_default(),
-			tx_fee,
-			(),
-			signature,
-		) {
-			log::error!("Failed to execute egress success: {:?}", err);
-		}
+		match contract_call_successful {
+			true =>
+				if let Err(err) = SolanaBroadcaster::egress_success(
+					pallet_cf_witnesser::RawOrigin::CurrentEpochWitnessThreshold.into(),
+					signature,
+					// Assign any owed fees to the current key.
+					SolanaThresholdSigner::active_epoch_key().map(|e| e.key).unwrap_or_default(),
+					tx_fee,
+					(),
+					signature,
+				) {
+					log::error!(
+						"Failed to execute egress success: TxOutId: {:?}, Error: {:?}",
+						signature,
+						err
+					)
+				},
+			false =>
+			// On CCM failure, we need to refund the user using their fallback info.
+				if let Some((broadcast_id, ccm_tx)) =
+					SolanaBroadcaster::pending_api_call_from_out_id(signature)
```

Comment:
Right. My suggestion would be the first one because in Solana an egress should be considered as successful once a transaction has been mined. Whether it reverted or not is not is not a factor in whether the egress was successful and all the actions should be the same.
Also in general and for future chains I think it can be useful to have a way to report that a transaction has been mined but has reverted on chain.

Reply:
Good point. THere's whole bunch of cleanup that needs to be done in the broadcast pallet...

Reply:
I'm a bit confused here. For all intends and purposes, the transaction should be counted as "successful", therefore we should just call the normal "egress_success" to handle the clean up right? the special to do the fallback logic is handled in the Solana hook. I don't see how the broadcast pallet needs to do anything different in the revert case.

Reply:
I've changed the logic so `egress_success` is always called. This ensures the calldata is cleared in all cases

Reply:
Thanks, i'll take a look. I think the naming in the broadcast pallet is confusing - conflates 'success' and 'failure' with 'complete' and 'aborted'.